Delete after revoke [#4109]

ynasser opened the pull request:
Implements #4109 

(This PR was previously in #4202)

ynasser commented:
Last commit message has detailed changes to make branch switching easier:
https://github.com/certbot/certbot/pull/4914/commits/8b4999caf91cf81477510cd5808d3dbb0e0ad8fd

Additionally, things I'd still like to do:
- Add docstrings/document parameter types in a way that Sphinx can understand.
- (Eventually) update the documentation in `using.rst` to reflect the new behaviour. I'll do that once we know what version of Certbot this will be in.

ynasser commented:
This is broken because of this: https://github.com/certbot/certbot/commit/93f094a86e4b137488ea1f83b7658ef5d7956592, which was introduced two weeks ago. Am updating the integration tests to reflect the new behaviour now.

ohemorange commented on certbot/cert_manager.py:
```diff
@@ -141,6 +142,93 @@ def update_certs_for_domain_matches(candidate_lineage, rv):
+                        if re.match("{0}[0-9]*.pem".format(filetype), f)]
+        # Using [0] here, so make sure to also check for overlapping archive dirs
+        # if you use this function.
+        return pattern[0]
```
it is expected that `len(pattern)` will often equal 0, meaning this will error out.

ynasser replied:
Good catch, thanks! I'm really surprised I missed this when I was testing earlier.

ynasser commented:
> At this point I think just unit tests, then we're good.

That was clear in your previous comment, I've just been tied up working on other stuff.

> Also, by this point I've forgotten - what did we say about the UX of this? Revoke should always automatically delete? Just when non-interactive?

I believe we originally decided that revoke will automatically delete when in non-interactive, otherwise we prompt.

ohemorange commented:
Ok, looks like it currently just always deletes.

ynasser commented:
> Ok, looks like it currently just always deletes.

Oops, sorry, I mixed up the prompts in my reply. What we decided was that we usually always delete, but if there are conflicting values for `--cert-name` and `--cert-path`, then we don't prompt in non-interactive mode and don't delete (which is taken care of via https://github.com/certbot/certbot/pull/4914/files#diff-0b4a662e2c7b44bd47156f3d976ed698R420).

ynasser commented:
@ohemorange I've written some unit and integration tests, and fixed a few issues along the way. Here's a summary of changes + tests since your last review to make reviewing this large PR a bit easier:

**Summary of non-testing changes**
- The most significant change is that the structure of `_acceptable_matches` and `acceptable_matches` in `cert_manager.cert_path_to_lineage` changed slightly.

**Summary of tests**
- `certbot.main._delete_if_appropriate`
  - No deleting happens if there are overlapping archive dirs
  - Only passing in `--certname` works
  - Only passing in `--cert-path` works
  - If run in non-interactive mode (with reasonable parameters), deleting happens.
  - If `--cert-name` and `--cert-path` are both used and they match, things work.
  - If `--cert-name` and `--cert-path` are mismatched in non-interactive mode
  - If `--cert-name` and `--cert-path`are mismatched in interactive mode, this can be resolved.
  - If neither `--cert-name` nor `--cert-path` is specified, then no deleting happens.
- `certbot.cert_manager.cert_path_to_lineage`
  - Basic match from a cert path to a cert name
  - The case where no match exists/isn't well defined
  - Test everything that can be matched against that's used in `acceptable_matches` actually works, e.g. (https://github.com/certbot/certbot/pull/4914/files#diff-900252b9d61ceaebc9038ba9af425b58R160)
- `certbot.cert_manager.match_and_check_overlaps`
  - Only test the main branches: a basic match is found, no matches found, and that `errors.OverlappingArchiveDirs is raised.`
  - The real test for detecting overlapping archive dirs was added to the integration tests in `tests/boulder_integration.sh`. I tried adding it as a unit test, but I felt like it was taking too much time and it took up a very large amount of space, so I switched to this.
- `certbot.storage.cert_path_for_cert_name`
  - Basic match from a cert path to a cert name
  - That `errors.CertStorageError` is raised if no match is found
- `tests/boulder_integration.sh`:
  - Checking that revoke deletes by default in practice (by checking that the revoked certs aren't listed in the output of `certbot certificates`)
  - Test that an error is raised if there are overlapping archive directories.

@ohemorange  You know a lot more about the specifics of certbot's certificate management than I do. Do you have suggestions for other things to test/check?

ynasser commented on certbot/cert_manager.py:
```diff
@@ -141,6 +142,110 @@ def update_certs_for_domain_matches(candidate_lineage, rv):
+
+    matched = _search_lineages(cli_config, find_matches, [], acceptable_matches)
+    if not matched:
+        raise errors.Error("No match found for cert-path {0}!".format(cli_config.cert_path))
```
No, but that makes a lot more sense.

ynasser commented:
> What is the rest of "If --cert-name and --cert-path are mismatched in non-interactive mode" supposed to say?

Nothing, I just missed the period. The unit test for that section was only to test that an error is raised if a user ends up in that state.

ynasser commented on certbot/main.py:
```diff
@@ -384,6 +387,76 @@ def _tos_cb(regr):
+
+    if config.certname and config.cert_path:
+        # first, check if certname and cert_path imply the same certs
+        cert_name_implied_cert_path = cert_manager.lineage_for_certname(config, config.certname)
```
Ah, I think the mismatch comes from a refactoring that happened a while ago. Both options don't seem great, but since 1 seems like it could really backfire if they've messed with the files in a way we don't predict during 1's  implementation, I'm going to go with 2 despite it being slower.

ynasser commented:
@ohemorange revoking by `--cert-name` is not possible, as `revoke` requires `--cert-path`. Did you mean to suggest something else?

ohemorange commented:
Then we should probably assert that `cert-path` has been specified at this point, instead: https://github.com/certbot/certbot/pull/4914/files#diff-0b4a662e2c7b44bd47156f3d976ed698R436

ynasser commented:
@ohemorange in a much earlier conversation, I thought we'd decided to support deleting by `--cert-name`, because there was a plan to _eventually_ having revoke support this. Can we just leave it in, with a note that when that happens someone should add the relevant integration test?

ohemorange commented:
Sure, I'm good with "add a note" as a solution here, especially if that note goes in the issue for "support revoking by `--cert-name` (which, does that exist?).

ynasser commented:
I added a note to the original issue: https://github.com/certbot/certbot/issues/4258#issuecomment-329311456

ynasser commented:
@ohemorange I've addressed your feedback. Here's a summary of the significant changes:
- In `certbot/main.py`:
  - Switched from `cert_manager.lineage_for_certname` -> `cert_manager.cert_path_to_lineage` and the other respective changes which needed to be made (based on this comment: https://github.com/certbot/certbot/pull/4914#discussion_r137903539).
  - Updated the tests in `certbot/tests/main_test.py` to reflect the above
- In `tests/boulder-integration.sh`:
  - Added a test to check that the correct error is raised if `--cert-name` and `--cert-path` don't match
  - Added a test for matching `--cert-name` and `--cert-path`
  - Didn't add the test for revoking by `--cert-name`, but someone else can when that feature is supported: https://github.com/certbot/certbot/issues/4258#issuecomment-329311456

Regarding your other suggestions, I've:
 - messed up the overlapping archive checking code and made sure tests failed (also, I'm pretty sure I've checked that most of the tests fail if the things they're supposed to be testing is broken)
 - manually tested that "mismatched `--cert-path` and `--cert-name` in interactive shows the correct certificates, for a variety of specified `--cert-path` options"

ynasser commented:
Updating the docstrings was in the TODO list further up this thread, but I totally forgot until you pointed that out. They should be updated now. I intentionally left out `human_readable_cert_info` because I know there's another PR that's waiting on this one that's going to change whatever I write.

ynasser commented:
@ohemorange is there anything else you'd like to see in this PR? I can update the docs in `using.rst` once I know what certbot version these changes are going to be released in.

ohemorange commented on certbot/main.py:
```diff
@@ -388,7 +388,15 @@ def _tos_cb(regr):
 
 def _delete_if_appropriate(config): # pylint: disable=too-many-locals,too-many-branches
-    """Does the user want to delete their now-revoked certs?"""
+    """Does the user want to delete their now-revoked certs? If run in non-interactive mode,
```
Doesn't it also happen automatically even if we're not in non-interactive mode?

ynasser replied:
Yes, I should re-phrase this. It was written a very, very long time ago.

ohemorange replied:
You know, are we sure this isn't the correct description, and the code is wrong? When did we decide to always just delete it? Didn't we decide to give an option, and have the default choice (ie when you just hit enter, implying that that's what happens in the non-interactive case) be to delete it?

ohemorange replied:
ah, found it https://github.com/certbot/certbot/pull/4202/files/6e07c333bf84a0869c82be2a5962cb2a3fb484c2#r115631846

ohemorange replied:
Given everything else about how the code currently is, what do you think?

ynasser replied:
Er, I'm a bit confused by your previous comment vs most recent ones. To be clear, which do you think is best?

Ultimately, I'm OK with either option, but this whole time I have been leaning towards at least giving users an option of saving their certs (as it might be useful for security auditing/retrospectives), even though I implemented deleting by default (because I thought that was the conclusion we came to as a result of some design discussions). It would only take a few hours to change it back to the previous behaviour.

ohemorange replied:
I really don't remember the previous design discussions we've had at this point, and the comment I put above was all I could find written down on the topic, which isn't much.

We also might want to worry about key pinning; people might want to keep their keys around even if they've revoked one cert.

If you're leaning towards asking, sounds good to me. It should really be only the one question, with default set to delete it, I think.

ynasser replied:
Ah, I was referencing an in-person meeting. And good point re: key pinning. I'll modify the behaviour/add the question in soon :)
